### DISCUSSION TITLE
How to start api in the laravel project?

### DISCUSSION THREAD

**Discussion by oraclecode318** [fattureincloud/api#481] [Q&A] [answered]
Hello, everyone.
I am going to implement the fattureincloud api in my laravel project.
I already created the account of "https://secure.fattureincloud.it/" and get the app_id, client_id, client_secret.
But I am not sure how to get the token and call the api in the Laravel backend.
I tested the getting token by generating the oauth.php in the phpmyadmin and successfully get the API token.

https://api-v2.fattureincloud.it/user/info

After that, I called above api in the postman, and it worked well.
But I am not sure how to start this in the Laravel project and how to manage the api token.
I already read the developer guide several times but not clear.

**Comment by valmoz** [comment]
Hello @quantumcoder482 
First, check if the OAuth 2.0 flow is the most appropriate for your use case. Check [here](https://developers.fattureincloud.it/docs/basics/authentication) if you need help deciding which Auth method to adopt.

Then, if you decide to use the OAuth 2.0 flow, you have two options:
- If you can install our [PHP SDK](https://developers.fattureincloud.it/docs/sdks/php-sdk), you can use our OAuth helpers to implement the OAuth 2.0 flow. Check [here](https://developers.fattureincloud.it/docs/authentication/code-flow/sdk-oauth-helper) for more info
- if you can't use our PHP SDK, you'll need to use an HTTP client such as [Guzzle](https://docs.guzzlephp.org/en/stable/). Check [here](https://developers.fattureincloud.it/docs/authentication/code-flow/vanilla-code) for more info

The OAuth2.0 flow is standard, so if you need more info you'll be able to find plenty of documentation about it even outside of our site. For example, [here](https://developers.fattureincloud.it/docs/general-knowledge/oauth2) you can find a generic description of the standard OAuth2.0 flow.
You can find a more detailed description of the Fatture in Cloud's version of the auth flow in the pages I've shared with you.

Please, let us know if you need further info about this topic.

Thanks
Mauro

**Reply by oraclecode318** [reply]
```
public function __construct()
    {
        parent::__construct();
        $this->pageTitle = 'app.menu.cloudInvoices';
        $this->pageIcon = 'ti-receipt';
        session()->put('lifetime', 86400);
        $appId = config('services.fatture_cloud.app_id');
        $clientId = config('services.fatture_cloud.client_id');
        $clientSecret = config('services.fatture_cloud.client_secret');
        $redirectUri = route('admin.cloud-invoices.index');
        $baseUri = "https://api-v2.fattureincloud.it";
        $oauth = new OAuth2AuthorizationCodeManager($clientId, $clientSecret, $redirectUri);

        if (isset($_SESSION['token'])) die('You already have an access token');

        if (!isset($_GET['code'])) {
            $url = $oauth->getAuthorizationUrl([Scope::ENTITY_SUPPLIERS_READ], "EXAMPLE_STATE");
            header('location: ' . $url);
        } else {
            $code = $_GET['code'];
            $obj = $oauth->fetchToken($code);
            if (!isset($obj->error)) {
                $_SESSION['token'] = $obj->getAccessToken(); //saving the oAuth access token in a session variable
                $_SESSION['refresh'] = $obj->getRefreshToken();
            }

            echo 'Token saved correctly in the session variable';
        }

        $this->middleware(function ($request, $next) {
            if (!in_array('invoicesPatients', $this->user->modules)) {
                abort(403);
            }
            return $next($request);
        });
    }
```

Above code is my Laravel contructor to get the access token.
It's working well on the localhost, but on the live site, it says redirectUri is not valid.
![Uploading image.png…]()

Would you like to let me know the reason of this?
and if I try to access the above  ```route('admin.cloud-invoices.index')```, it requires the login https://secure.fattureincloud.it/
I think some thing went wrong. Users of my site don't know the credential.

**Reply by valmoz** [reply, marked as answer]
Hello @quantumcoder482 
the redirect URL must be exactly equal to the one you inserted on the App inside the Fatture in Cloud's "Sviluppatore" section.
You left the localhost redirect URL in the App config.

It is probably better to create multiple Fatture in Cloud apps, one for each environment you use.

Thanks
Mauro

**Reply by oraclecode318** [reply]
Thank you for your answer.

**Reply by valmoz** [reply]
@quantumcoder482 you're welcome
